### DISCUSSION TITLE
Separating OPA installation from policy loading

### DISCUSSION THREAD

**Discussion by rohithdv** [fybrik/fybrik#420] [Developers]
We want to separate OPA installation from the way we load policies to OPA in mesh for data. This discussion deals with approaches to do the same.
We have implemented the below approach :
Assume we are working in the namespace <OPA_NAMESPACE>
Step 1. Delete all existing resources:

```
kubectl delete -f policy-editor.yaml
kubectl delete -f policy-editor-rolebinding.yaml
kubectl delete -f opa.yaml
kubectl delete configmap opa-json
kubectl delete configmap opa-policy
```

Step 2, Create Role and Rolebinding resources and associate them with the <OPA_NAMESPACE>.
`kubectl create -f policy-editor.yaml`

```
kind: Role
apiVersion: rbac.authorization.k8s.io/v1
metadata:
namespace:  <OPA_NAMESPACE>.
name: policy-editor
rules:
apiGroups: [""]
resources: ["configmaps"]
verbs: ["get", "list", "watch", "update", "patch"]
```

`kubectl create -f policy-editor-rolebinding.yaml`

```
kind: RoleBinding
apiVersion: rbac.authorization.k8s.io/v1
metadata:
name: policy-editor
namespace:  <OPA_NAMESPACE>.
roleRef:
kind: Role
name: policy-editor
apiGroup: rbac.authorization.k8s.io
subjects:
kind: Group
name: system:serviceaccounts:<OPA_NAMESPACE>.
apiGroup: rbac.authorization.k8s.io
```

Step 3. Deploy plain vanilla OPA with kube-mgmt using the command  kubectl create -f opa.yaml. The spec of the opa.yaml is given below.
```
    spec:
      containers:
      - name: opa
        image: openpolicyagent/opa:latest
        args:
        - "run"
        - "--server"
        - "--log-level=debug"
        - "--set=decision_logs.console=true"
        ports:
        - name: http
          containerPort: 8181
      - name: kube-mgmt
        image: openpolicyagent/kube-mgmt:0.11
        args:
        - --require-policy-label
        - --policies=irltest1
        - --enable-data
```

Step 4. Create a Kustomization resource with only the data related files (json data)
```
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
generatorOptions:
  disableNameSuffixHash: true
configMapGenerator:
- name: opa-json
  files:
  - taxonomies.json
  - medical_taxonomies.json

```

Step 5. Load the data to OPA using kube-mgmt labelling approach (https://github.com/open-policy-agent/kube-mgmt)
```
kubectl kustomize ./ |kubectl label -f- --dry-run=client -o yaml --local openpolicyagent.org/data=opa > newjson.yaml
kubectl create -f newjson.yaml
```

Step 6. Create a Kustomization resource with only the policy related files (rego data)
```
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
generatorOptions:
  disableNameSuffixHash: true
configMapGenerator:
- name: opa-policy
  files:
  - data_policies/action_struct.rego
  - data_policies/auditing_struct.rego
  - data_policies/helper_functions.rego
  - data_policies/input_reader.rego
  - data_policies/taxonomies_unification.rego
  - data_policies/verify_correct_input.rego
  - sample_policies.rego
  - user_policies.rego
```

Step 7. Load the policies to OPA  using kube-mgmt labelling approach (https://github.com/open-policy-agent/kube-mgmt)
```
kubectl kustomize ./ |kubectl label -f- --dry-run=client -o yaml --local openpolicyagent.org/policy=rego > newpolicy.yaml
kubectl create -f newpolicy.yaml
```

One thing to discuss here is how to access helper data (json data) in policy files. One way to do this is given in [taxonomies_unification.rego](https://github.com/rohithdv/the-mesh-for-data/blob/59259df6bcdb3cdd4d5f5030fe7ce7effcf88845/third_party/opa/data-and-policies/data_policies/taxonomies_unification.rego#L8).

As seen in the above taxonomies_unification.rego file, there is a hardcoding of namespace to access data in the policy file. We can discuss if there are better ways to do the above.

To summarise,
Step 1 -2  deals with initialisation of the namespace and also setting permissions to read configmaps to the namespace.

Step 3 deals with installation of OPA with kube-mgmt as a sidecar

Step 4-7 deals with loading policies (and any associated data) to OPA. Note that this step could be done at different instants of time for different policies.

Is this approach ok for delineating OPA installation with policy loading ? Please let me know your feedback. Thanks .
@roee88 @ronenkat @simanadler @ritwikc @shlomitk1

**Comment by roee88** [comment]
Please use 
```
formatting
```

**Reply by rohithdv** [reply]
Actually I have used formatting when I created this post in desktop. Somehow the formatting goes away in GitHub mobile app.

**Reply by rohithdv** [reply]
I have recreated the post. Hopefully the formatting will show now.

**Comment by roee88** [comment]
I only looked briefly so it might be partial feedback, but general comments:
1. You are mixing files that your current implementation requires (like action_struct.rego) with files that the user can add. Nothing forces it to be in the same configmap so you can fully separate deployment from actual user input. 
2. The use of kustomize is just an example. The user can use a different method to create these configmaps from rego. Also there is no assumption that all user policies are in the same configmap. 
3. Hardcoding the namespace is an issue but we want to move away from handling taxonomy like that so it would eventually go away. I think we can leave it as m4d-system for now unless you can find an easy way to detect it.
4. Not described here but needs to be done: use a predefined and fixed package name in all policies (Just like `httpapi.authz` is used for HTTP policies) and then always query using the data API of OPA. This eliminates specifying which policy name to query in m4d-config (currently set to "sample_policies").

**Reply by rohithdv** [reply]
Thanks @roee88 . To address (1), I have re-organized the dir structure to handle your comment. Can you please check [here](https://github.com/rohithdv/the-mesh-for-data/tree/opa_use_kube_mgmt3/third_party/opa/data-and-policies) and let me know if this is what you intended ? The overall script to handle opa installation and data and policy loading is given [here](https://github.com/rohithdv/the-mesh-for-data/blob/opa_use_kube_mgmt3/third_party/opa/delete_and_redeploy_opa.sh) for your reference. Thanks.

**Reply by rohithdv** [reply]
To address (2), I think the user can use any way to create the configmaps. kube-mgmt will be agnostic to it.

**Reply by rohithdv** [reply]
To address (3), I have followed the documentation of kube-mgmt (https://github.com/open-policy-agent/kube-mgmt). Please check the JSON loading section in the README. I will try to check if there are any better ways to handle this. Thanks.

**Reply by rohithdv** [reply]
For Point (4), suppose we made a common package across all policies, say meshfordata.authz, now in the current user_policies.rego there is a transform rule and in the current sample_policies.rego there is a transform rule. Once we have a common policy name, will it not be difficult to distinguish which transform rule we need to check in the code ?
